src/main/scala/com/normation/cfclerk/domain/AixPasswordHashAlgo.scala

```diff
@@ -0,0 +1,366 @@
+    for {
+      skf <- getSecretKeFactory(sha)
+    } yield {
+      doSsha(sha, skf)(pwd, salt, cost)
+    }
+  }
+
+  /*
+   * Create an AIX {ssha1} hash string from pwd. Optionaly, specify the
+   * salt string to use (default to a random 16 chars string), and the "num_cost"
+   * (defaults to 10, i.e 1024 iterations).
+   *
+   * If PBKDF2WithHmacSHA1 crypto algorithm is not available in the current JVM,
+   * revert back to {smd5}.
+   */
+  def ssha1(pwd: String, salt: Option[String] = None, cost: Int = 10): String = {
+    ssha1impl(pwd, salt, cost)
+  }
+
+  /*
+   * Create an AIX {ssha256} hash string from pwd. Optionaly, specify the
+   * salt string to use (default to a random 16 chars string), and the "num_cost"
+   * (defaults to 10, i.e 1024 iterations).
+   *
+   * If PBKDF2WithHmacSHA256 crypto algorithm is not available in the current JVM,
+   * revert back to {ssha1}.
+   */
+  def ssha256(pwd: String, salt: Option[String] = None, cost: Int = 10): String = {
+    ssha256impl(pwd, salt, cost)
+  }
+
+  /*
+   * Create an AIX {ssha256} hash string from pwd. Optionaly, specify the
+   * salt string to use (default to a random 16 chars string), and the "num_cost"
+   * (defaults to 10, i.e 1024 iterations).
+   *
+   * If PBKDF2WithHmacSHA512 crypto algorithm is not available in the current JVM,
+   * revert back to {ssha1}.
+   */
+  def ssha512(pwd: String, salt: Option[String] = None, cost: Int = 10): String = {
+    ssha512impl(pwd, salt, cost)
+  }
+
+
+  /////
+  /////          Implementation details
+  ///// (if you are looking to understand how AIX hashes its
+  /////  passwords, that the part of interest)
+  /////
+
+  protected[domain] final def getSecretKeFactory(sha: ShaSpec): Box[SecretKeyFactory] = {
+    try {
+      Full(SecretKeyFactory.getInstance(s"PBKDF2WithHmac${sha.name}"))
+    } catch {
+      case ex: NoSuchAlgorithmException =>
+        Failure(s"Your current Java installation does not support PBKDF2WithHmac${sha.name} algorithm, " +
+            "which is necessary for {ssha256} hash")
+    }
+  }
+
+  /////
+  ///// Initialize the correct instance of hashing algo
+  ///// based on the available secret key factory on the
+  ///// jvm. The check is done at most one time by jvm session.
+  /////
+
+  /// ssha1 revert to smd5 - but all post-java6 JVM should be ok
+  private[this] final lazy val ssha1impl = getSecretKeFactory(ShaSpec.SHA1) match {
+    case Full(skf) => doSsha(ShaSpec.SHA1, skf) _
+    case e:Failure =>
+      // this should not happen, because PBKDF2WithHmacSHA1 is
+      // in standard Java since Java6. But who knows..
+      // Fallback to md5 hash.
+      logger.error("Your current Java installation does not support PBKDF2WithHmacSHA1 algorithm, " +
+          "which is necessary for {ssha1} hash. Falling back to {smd5} hashing scheme")
+
+      (pwd: String, salt: Option[String], cost: Int) => smd5(pwd, salt)
```

Comment:
i'm not 100% sure of the syntax there. does that mean that it take as parameters (pwd: String, salt: Option[String], cost: Int) ?
shouldn't it be a def then rather than a val?